_posts/2024-03-20-#29221.md

```diff
@@ -0,0 +1,198 @@
+
+## How arithmetic works with #29221
+
+Three key differences exist in how 64-bit opcodes function compared to their previous counterparts:
+
+1. **Enhanced Precision**: They support 64 bits of precision, enabling more accurate arithmetic operations.
+
+2. **Error Handling Capability**: These opcodes provide error handling by pushing either true or false onto the stack, depending on whether the operation succeeds or fails.
+
+3. **Standardized Encoding**: They utilize a consistent fixed-length 8-byte number encoding format, aligning with conventions elsewhere in the Bitcoin codebase, such as in `CTxOut::nValue`.
+
+As an illustration of the third difference, consider the encoding of 1,000 BTC. It would now be represented in the same format as seen on a block explorer (`0x00e8764817000000`) rather than `0x00e8764817` which is the CScriptNum encoding.
+
+### Example: Adding 1,000 BTC together with OP_ADD64
+
+Here's the same example from above with `OP_ADD64` rather than `OP_ADD` with our new little endian encoding format rather than `CScriptNum`:
+```
+0x000e876481700000 0x000e876481700000 OP_ADD64 OP_DROP 0x001d0ed902e00000 OP_EQUAL
+```
+
+```
+[Stack: ] --(0x00e8764817000000)--> [Stack: 0x00e8764817000000]
+          --(0x00e8764817000000)--> [Stack: 0x00e8764817000000, 0x00e8764817000000]
+          --(OP_ADD64)--> [Stack: 0x01d0ed902e000000, true]
+          --(OP_DROP)--> [Stack: 0x01d0ed902e000000]
+          --(0x01d0ed902e000000)--> [Stack: 0x01d0ed902e000000, 0x01d0ed902e000000]
+          --(OP_EQUAL)--> [Stack: true]
+```
+
+Explanation:
+
+1. The initial state of the stack is empty.
+2. `0x00e8764817000000` pushes the hexadecimal value `0x00e8764817000000` onto the stack (representing 100,000,000,000 satoshis).
+3. Another instance of `0x00e8764817000000` is pushed onto the stack.
+4. `OP_ADD64` attempts to pop the top two elements (`0x00e8764817000000` and `0x00e8764817000000`) to add them. The correct result of the addition `0x01d0ed902e000000` (representing 200,000,000,000 satoshis) is pushed onto the stack first, followed by `true`, indicating that the arithmetic executed correctly.
+5. `OP_DROP` drops the `true` pushed onto the stack by OP_ADD64 indicating the arithmetic operation was successfull.
+6. `0x001d0ed902e00000` pushes the hexadecimal value `0x001d0ed902e00000` onto the stack (representing 200,000,000,000 satoshis).
+7. `OP_EQUAL` compares the two top stack values `0x001d0ed902e00000` and pushes `true` onto the stack
+
+## Design questions
+
+### Signed vs unsigned arithmetic
+
+Much of the implementation uses code from the [elements blockchain](https://github.com/ElementsProject/elements/). In elements they implemented new arithmetic opcodes as fixed size 64 bit signed integers.
+Do we have a use case for using signed math rather than unsigned math? My satoshi example would work with unsigned math (outputs can't have negative value) even though sats are encoded
```

Comment:
```suggestion
Do we have a use case for using signed math rather than unsigned math? The satoshi example would work with unsigned math (outputs can't have negative value) even though sats are encoded
```